### ISSUE TITLE
questions about dropout_add_layer_norm

### ISSUE THREAD

**Issue by yingtongxiong** [Dao-AILab/flash-attention#180]
Hi, I have some questions about dropout_add_layer_norm. I implements the dropout_add_layer_norm by using torch.nn.Dropout and torch.nn.Layernorm. When I want to know whether my results is equal to the flash-attn version, I find out that the results are different even if I have set the random seed. Therefore, do you have any solutions to make the dropout identical?

These are my code to set the random seed:

torch.manual_seed(0)
torch.cuda.manual_seed(0)
torch.cuda.manual_seed_all(0)
torch.backends.cudnn.deterministic = True

However, the dropout mask generated is not same between flash-attn and pytorch version. Do you have any suggestions to solve it? Thank you.

**Comment by tridao** [comment]
There's no way to get the same dropout mask from two different dropout implementations.
As an example, there are at least 2 correct ways to implement dropout with probability 0.4: generate uniformly random numbers between [0, 1), and then mask out numbers < 0.4, or mask out numbers >= 0.6. They will not have the same dropout mask even from the same seed.
```
import torch
a = torch.randn(3, 3, device='cuda')
torch.manual_seed(0)
b = (torch.rand_like(a) < 0.4) * a / (1 - 0.4)
torch.manual_seed(0)
c = (torch.rand_like(a) >= 0.6) * a / (1 - 0.4)
# b and c will have different dropout patterns
```
You can look at the way we [test](https://github.com/HazyResearch/flash-attention/blob/3da42d24b1678b3ed930ab86e801b4ee851af1d2/tests/ops/test_dropout_layer_norm.py#L110) dropout_add_layer_norm: get the dropout mask from the flash-attn version, then explicitly mask out those entries in the pytorch implementation, and check that the outputs are the same (up to numerical precision).

**Comment by yingtongxiong** [comment]
Okay, thank you, I have tested the solution you mentioned above. Besides, I want to test the performance of two different implementation, which needs the input, model weights, dropout mask and something else same. So, do you have any suggestions about conducting the performance experiments?

**Comment by tridao** [comment]
By performance do you mean speed? Then you can just run those operations and measure the time, they don't have to produce the same exact output.

If by performance you mean model quality (i.e. accuracy or perplexity), then the simplest is to train models with two different implementations and look at the metrics (accuracy, perplexity).
